### ISSUE TITLE
Book position syncing across devices

### ISSUE THREAD

**Issue by major-laslo** [Ph1b/MaterialAudiobookPlayer#295]
It would be nice to have book position syncing across devices the way pocket cast does, so I can switch from phone to tablet and have the app simply know where I was. 

I guess this would require (optional) authentication (with google account?) and a simple backend (with google app engine maybe?) and some way to notify the other device (GCM?), so I see how this feature would not be a small undertaking...so a 'down the line' feature?

It's not my area of expertise but if you agree with the general idea, I'd be happy to try and contribute to building it.

**Comment by Ph1b** [comment]
Reference #65 

The problem with this is the way how book adding works. In principle I think this would be a good thing to have.
But the first major problem: How do we identify the books as the same? Now we see them as equal if they exactly match (every chapter is the same). 

But yeah, if you like to start working on this, sure! Keep software licenses in mind ;-)

**Comment by major-laslo** [comment]
Yep, I wanted to check with you before putting significant time into it, since there is a lot of homework to be done before even starting...(settling on book ids, user ids, backend/syncing solution). If the feature is kept as opt-in, and the models kept simple, it should always fall within the free usage quota of Google App Engine.

Another (and probably simpler) option is Firebase, which would handle auth, db and propagating changes to all clients and persistence...also with a generous free quota.

I'll look into things and study your code a bit, and come back with a clear proposal in a week or two max.

**Comment by major-laslo** [comment]
Ok, so I played around a bit with Firebase with authentication and google login, I believe it would be a pretty good and simple way to implement this with the absolute minimum effort to the user (and development). License of Firebase looks ok for this https://www.firebase.com/terms/terms-of-service.html and the only other thing we'd need is the Google Play Services (Plus API) which is also free.

The free usage quotas of Firebase 'should' cover the application's needs, only bottleneck might be the 100 concurrent connections, but with a fallback mechanism it shouldn't be an issue. The data that needs to be synced and stored should be negligibly small, I believe.

I made a small todo app as a proof of concept, to see what the syncing and login would consist of, it's pretty dirty but https://bitbucket.org/major_laslo/todocloudmeow/ if you want to glance at it. (can't use it though as it's tied to my debug key)

All this could be made completely opt-in, so anyone who doesn't want to use it or doesn't have google play services notices no difference.

As you pointed out, the first major issue that would need solving is how to identify a book uniquely but consistently across devices...maybe generate a hash based on the name/chapters? If you have some pointers, I'll start looking into it and trying out things in the player.

**Comment by Ph1b** [comment]
I am not sure about the connection bottleneck.

At the moment we have an active userbase of about 15k, growing. And that includes only the google play downloads, not the FDroid or manual ones. I would assume that this breaks the limit.

**Comment by major-laslo** [comment]
It only applies during the moment of syncing I believe, which could/should be only when a book is paused or the app is getting stopped/destroyed. I doubt that it would often happen that 100 users are triggering a sync at the same second. And I think Firebase would fall back and retry on its own if it was unable to sync.

**Comment by major-laslo** [comment]
Do you still believe that it would not be viable, and some other solution should be investigated? Google App Engine could be an option but it would be a bit more involved as we'd need to take care of the DB, and GCM...but all 'free' options will have some limitations (and pretty same-ish ones), and to me it looks like Firebase takes care a lot of things that we need out of the box.

Either way, I'll not try to force something on you that you might not be comfortable with, just though this could be a pretty 'cheap' and seamless way to implement a useful feature.

If you do think it's worth a try and would like me to start implementing on a fork, we'll first need to figure out a consistent and reliable way to identify the books - and this is reaching into one of the core parts of the application so it will have to be something you are comfortable with.

**Comment by Thomas-O** [comment]
As a quick rough-and-ready solution, what about storing the positions in a plain text file in the audiobooks folder? Something which the user can easily sync using BtSync, Syncthing or OwnCloud.

**Comment by asommer70** [comment]
I love Material Player too.  Picking up an audio book/podcast exactly where I left off is awesome.  

I had the idea to allow the user to enter a URL which they can then sync to using a simple Node backend.  Like this [one](https://github.com/asommer70/audiopila-node).  The idea is to host the server on your local home network and sync multiple devices through that.  There currently isn't any user authentication, but there could be.  That way you don't have to worry about 3rd party cloud quotas, integrations, etc.  Can also donwload/upload audio files to the server.

Another ideas is to sync the data over Bluetooth, but that get's more complicated and might be a feature for later on.

What do you think about a self-hosted audio data server?

**Comment by rejhgadellaa** [comment]
I'm afraid i'm not going to be able to help develop this but +1 because i'd really like this feature

**Comment by asommer70** [comment]
@rejhgadellaa that's cool, it's on my list of ToDos.  So maybe some day :-)

**Comment by Wizek** [comment]
+1 for this from me too! 

@major-laslo, how about the following for simplicity?

We could store the positions indexed by folder name for the book. 

```json
{ 
  "Book_folder_name_1": "00h10m21s",
  "Book_folder_name_2": "00h00m00s"
}
```

That way, if I sync to a device that doesn't have a certain folder locally, it would still show up in the app GUI, perhaps grayed out, perhaps even saying so why when it is tapped on: "This book doesn't seem to exist on this device. If you think this is an error, please make sure that the folder names match exactly across devices."

**Comment by freman** [comment]
> We could store the positions indexed by folder name for the book.

That's what I was poking around trying to do, but I'm not an android dev, never used Kotlin, and am entirely confused by all this @inject stuff

I have a working sync setup with a locally hosted service, but I can't figure out how to actually get or put the data within the rest of the app 😢
